statefun-e2e-tests/statefun-smoke-e2e-embedded/src/test/java/org/apache/flink/statefun/e2e/smoke/embedded/EmbeddedSmokeHarnessMigrationTest.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.statefun.e2e.smoke.embedded;
+
+import static org.apache.flink.statefun.e2e.smoke.SmokeRunner.awaitVerificationSuccess;
+
+import java.io.File;
+import java.net.URI;
+import java.util.Arrays;
+import java.util.Collection;
+import org.apache.commons.io.FileUtils;
+import org.apache.flink.core.execution.JobClient;
+import org.apache.flink.statefun.e2e.smoke.SimpleVerificationServer;
+import org.apache.flink.statefun.e2e.smoke.SmokeRunnerParameters;
+import org.apache.flink.statefun.flink.harness.Harness;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@RunWith(Parameterized.class)
+public class EmbeddedSmokeHarnessMigrationTest {
```

Comment:
Thanks for adding a migration test!
This test would catch any internal migration issues with the core statefun job graph.

Perhaps we need to followup with a separate e2e migration test, that runs within its own migration profile, that includes at least a Kafka ingress/egress pair.
It can be completely based remote functions, without any verification/failure injection aspects to it, and this probably should be based on `testcontainers` instead of the harness.

having test containers will allows the test to create a save point in an arbitrary previous version (say 3.1.1) 
based on the previously released official docker images, and load it with the current version.
The test would be bit more time consuming, and that's why I would recommend that to run under a different profile.

If this is something you find reasonable, then I'll create a followup JIRA issue.

Reply:
I think this sounds like a good idea. Having these test to ensure that nothing is breaking unintentionally is a good improvement. I will remove the commit that introduced the harness test and we continue based on the migration test JIRA ticket you'll create.